test/util/chainrules.jl

```diff
@@ -0,0 +1,166 @@
+        test_rrule(_map, sum, x; check_inferred=false)
+        test_rrule(_map, x->map(sin, x), x; check_inferred=false)
+        test_rrule(_map, x -> 2.0 * x, x; check_inferred=false)
+        test_rrule(ZygoteRuleConfig(), (x,a)-> _map(x -> x * a, x), x, 2.0; check_inferred=false, rrule_f=rrule_via_ad)
+    end
+    @testset "_map(f, x1::Fill, x2::Fill)" begin
+        x1 = Fill(randn(3, 4), 3)
+        x2 = Fill(randn(3, 4), 3)
+
+        @test _map(+, x1, x2) == _map(+, collect(x1), collect(x2))
+        test_rrule(_map, +, x1, x2; check_inferred=true)
+
+        fsin(x, y) = sin.(x .* y)
+        test_rrule(_map, fsin, x1, x2; check_inferred=false)
+
+        foo(a, x1, x2) = _map((z1, z2) -> a * sin.(z1 .* z2), x1, x2)
+        test_rrule(ZygoteRuleConfig(), foo, randn(), x1, x2; check_inferred=false, rrule_f=rrule_via_ad)
+    end
+    # @testset "$N, $T" for N in [1, 2, 3], T in [Float32, Float64]
+
+    #     rng = MersenneTwister(123456)
+
+    #     # Do dense stuff.
+    #     S_ = randn(rng, T, N, N)
+    #     S = S_ * S_' + I
+    #     C = cholesky(S)
+    #     Ss = SMatrix{N, N, T}(S)
+    #     Cs = cholesky(Ss)
+
+    #     @testset "cholesky" begin
+    #         C_fwd, pb = Zygote.pullback(cholesky, Symmetric(S))
+    #         Cs_fwd, pbs = Zygote.pullback(cholesky, Symmetric(Ss))
+
+    #         @test eltype(C_fwd) == T
+    #         @test eltype(Cs_fwd) == T
+
+    #         ΔC = randn(rng, T, N, N)
+    #         ΔCs = SMatrix{N, N, T}(ΔC)
+
+    #         @test C.U ≈ Cs.U
+    #         @test Cs.U ≈ Cs_fwd.U
+
+    #         ΔS, = pb((factors=ΔC, ))
+    #         ΔSs, = pbs((factors=ΔCs, ))
+
+    #         @test ΔS ≈ ΔSs.data
+    #         @test eltype(ΔS) == T
+    #         @test eltype(ΔSs.data) == T
+
+    #         @test allocs(@benchmark(cholesky(Symmetric($Ss)); samples=1, evals=1)) == 0
+    #         @test allocs(@benchmark(Zygote._pullback($(Context()), cholesky, Symmetric($Ss)); samples=1, evals=1)) == 0
+    #         @test allocs(@benchmark($pbs((factors=$ΔCs,)); samples=1, evals=1)) == 0
+    #     end
+    #     @testset "logdet" begin
+    #         @test logdet(Cs) ≈ logdet(C)
+    #         C_fwd, pb = logdet_pullback(C)
+    #         Cs_fwd, pbs = logdet_pullback(Cs)
+
+    #         @test eltype(C_fwd) == T
+    #         @test eltype(Cs_fwd) == T
+
+    #         @test logdet(Cs) ≈ Cs_fwd
+
+    #         Δ = randn(rng, T)
+    #         ΔC = first(pb(Δ)).factors
+    #         ΔCs = first(pbs(Δ)).factors
+
+    #         @test ΔC ≈ ΔCs
+    #         @test eltype(ΔC) == T
+    #         @test eltype(ΔCs) == T
+
+    #         @test allocs(@benchmark(logdet($Cs); samples=1, evals=1)) == 0
+    #         @test allocs(@benchmark(logdet_pullback($Cs); samples=1, evals=1)) == 0
+    #         @test allocs(@benchmark($pbs($Δ); samples=1, evals=1)) == 0
+    #     end
+    # end
+    @testset "StructArray" begin
+        a = randn(5)
+        b = rand(5)
+        test_rrule(StructArray, (a, b))
```

Comment:
I have a weird error here where the tangent obtained with FiniteDifferences.jl seems quite random (some permutations and perturbations?).
@willtebbutt Could you have a look if the `rrule` seems correct?

Reply:
I ended up commneting out the tests because they seem  to come from FiniteDifferences.jl